crates/goose-server/src/routes/agent.rs

```diff
@@ -342,6 +357,146 @@ async fn update_session_config(
+        fn normalize_like_backend(input: &str) -> String {
+            let mut result = String::with_capacity(input.len());
+            for c in input.chars() {
+                match c {
+                    c if c.is_ascii_alphanumeric() || c == '_' || c == '-' => result.push(c),
+                    c if c.is_whitespace() => { /* skip */ }
+                    _ => result.push('_'),
+                }
+            }
+            result.to_lowercase()
+        }
+
+        // Split incoming tool name into prefix and tool part if provided
+        let (incoming_prefix_opt, incoming_tool_part) = match resolved_tool_name.split_once("__") {
+            Some((p, t)) => (Some(p.to_string()), t.to_string()),
+            None => (None, resolved_tool_name.clone()),
+        };
+
+        // Fetch registered extensions
+        let registered_extensions: Vec<String> = {
+            let extension_manager = agent.extension_manager.read().await;
+            extension_manager
+                .list_extensions()
+                .await
+                .unwrap_or_default()
+        };
+
+        // Try to map by prefix if provided
+        let mut remapped = false;
+        if let Some(incoming_prefix) = incoming_prefix_opt.as_ref() {
+            let norm = normalize_like_backend(incoming_prefix);
+            let norm_trim = norm.trim_matches('_');
+            for ext in &registered_extensions {
+                let ext_trim = ext.trim_matches('_');
+                if norm_trim == ext_trim || norm_trim == ext.as_str() {
+                    let candidate = format!("{}__{}", ext, incoming_tool_part);
+                    if available_tools.iter().any(|t| t.name == candidate) {
+                        resolved_tool_name = candidate;
+                        remapped = true;
+                        break;
+                    }
+                }
+            }
+        }
+
+        // If still not found, try suffix match (any tool ending with __<name>)
+        if !remapped {
+            let suffix = format!("__{}", resolved_tool_name);
+            if let Some(candidate) = available_tools
+                .iter()
+                .find(|t| t.name.ends_with(&suffix))
+                .map(|t| t.name.clone().to_string())
+            {
+                resolved_tool_name = candidate;
+            }
+        }
+    }
+
+    // Create a tool call with the resolved name
+    let tool_call = mcp_core::tool::ToolCall::new(resolved_tool_name, payload.arguments);
+
+    // Generate a unique request ID using timestamp
+    let request_id = format!("tool_{}", chrono::Utc::now().timestamp_millis());
+
+    // Create a cancellation token for the tool execution
+    let cancellation_token = tokio_util::sync::CancellationToken::new();
+
+    // Dispatch tool call
+
+    // Execute the tool
+    let (_returned_id, result) = agent
+        .dispatch_tool_call(tool_call, request_id, Some(cancellation_token))
```

Comment:
this just doesn't work. there's a reason toolcalling in the agent is a complex process that handles corner cases, handles streaming etc